consensus/spos/invalidSignersCache.go

```diff
@@ -0,0 +1,131 @@
+package spos
+
+import (
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/hashing"
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	"github.com/multiversx/mx-chain-go/consensus"
+	"github.com/multiversx/mx-chain-go/p2p"
+)
+
+// ArgInvalidSignersCache defines the DTO used to create a new instance of invalidSignersCache
+type ArgInvalidSignersCache struct {
+	Hasher         hashing.Hasher
+	SigningHandler p2p.P2PSigningHandler
+	Marshaller     marshal.Marshalizer
+}
+
+type invalidSignersCache struct {
+	sync.RWMutex
+	invalidSignersHashesMap    map[string]struct{}
+	invalidSignersForHeaderMap map[string]map[string]struct{}
```

Comment:
So, for each header, we retain the public keys of the "bad" signers (as a map, for easy lookup), correct?